### PR TITLE
Add Gaussian NLL loss

### LINKED ISSUE #48520: Gaussian NLL loss
## 🚀 Feature
Gaussian negative log-likelihood loss, similar to issue #1774 (and solution pull #1779) 

## Motivation

The homoscedastic Gaussian loss is described in Equation 1 of this [paper](https://arxiv.org/pdf/1612.01474.pdf). The heteroscedastic version in Equation 2 [here](https://arxiv.org/pdf/2011.02838.pdf)  (ignoring the final anchoring loss term). These are both key to the uncertainty quantification techniques described. 

## Pitch

I'm happy to implement this, using the template of pull #1779. The implementation will allow for both homoscedastic and heteroscedastic losses.  

## Alternatives

An alternative would be to instantiate a Gaussian (https://pytorch.org/docs/stable/distributions.html#normal) and evaluate the log of this. However, this seems wasteful given a new Gaussian would be instantiated for every new function call for the best case (homoscedastic), and for every element of the output-target pairs in the worst case (heteroscedastic). 

## Additional context

Definitions: [homo/heteroscedasticity](https://en.wikipedia.org/wiki/Homoscedasticity) (wiki)

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by nailimixaM**
Fixes #48520.

cc @albanD (This is a retry of PR #49089)

### CODE REVIEW

File: torch/nn/functional.py
Comment (albanD) [inline torch/nn/functional.py:2357]: nit: Isn't it better to use clamp here instead of add? That way you won't modify values that don't need to be modified.

You can even do that in `torch.no_grad()` if you think that the clamping should be ignored wrt autograd.

File: torch/nn/functional.py
Comment (albanD) [inline torch/nn/functional.py:2319]: 1e-8 doesn't sound great here.
For single precision float, you want to use 1e-6 in general to avoid any issue. And for double precision you want something much smaller.

I think that 1e-6 and clamping (see other comment below) should be better.

File: torch/nn/modules/loss.py
Comment (albanD) [inline torch/nn/modules/loss.py:305]: sum?

File: torch/nn/modules/loss.py
Comment (albanD) [inline torch/nn/modules/loss.py:307]: nit: capitalisation?

File: torch/nn/modules/loss.py
Comment (albanD) [inline torch/nn/modules/loss.py:344]: Same comment about the default eps

File: torch/nn/functional.py
Comment (albanD) [inline torch/nn/functional.py:2319]: nit: this might be better to make these kwargs-only to ensure people use the api properly by doing:
```
 def gaussian_nll_loss(input, target, var, *, eps=1e-8, full=True, reduction='mean'): 
```

Same for the nn.Module version.

**Review by albanD — COMMENTED** [review]
The change looks good.
Some small comments about defaults.

Also did you manage to build the doc locally?

File: torch/nn/functional.py
Comment (nailimixaM) [reply torch/nn/functional.py:2357]: `var` is the output of an NN here so requires grad, so I can't use clamp as it'll kill the gradients at the clamped values.

File: torch/nn/functional.py
Comment (nailimixaM) [reply torch/nn/functional.py:2319]: See my comment about clamping. Should I still use 1e-6 in this case?

File: torch/nn/modules/loss.py
Comment (nailimixaM) [reply torch/nn/modules/loss.py:305]: Using sim to mean the target is Gaussian distributed with mean 'input' and variance 'var'.

File: torch/nn/modules/loss.py
Comment (nailimixaM) [reply torch/nn/modules/loss.py:307]: Yup, thanks!

File: torch/nn/functional.py
Comment (nailimixaM) [reply torch/nn/functional.py:2319]: Good idea, thanks.

**Comment by nailimixaM** [comment]
@albanD Still working on building the docs locally. Will confirm with screenshots when done.

**Comment by nailimixaM** [comment]
@albanD As promised, see snap of doc entry below. I appreciate your comments on this too.
doc

File: torch/nn/functional.py
Comment (albanD) [inline torch/nn/functional.py:2328]: This is not in your screenshot, but isn't this line missing an extra indent? Other docstring in this file seem to have them.

File: torch/nn/functional.py
Comment (albanD) [reply torch/nn/functional.py:2357]: That is why I mentioned above that you don't have to do that in a differentiable way and you can do:
```python
# Avoid changing the argument given by the user inplace
var = var.clone()
with torch.no_grad():
  var.clamp_(min=eps)
```

File: torch/nn/modules/loss.py
Comment (albanD) [inline torch/nn/modules/loss.py:308]: I feel like it would be nice to have only a single formula here.  In particular, isn't this one a special case of the one below?
Similarly for the third one, can you just say that if var is a scalar, it is broadcasted to be the same size as the  input?

**Review by albanD — COMMENTED** [review]
Thanks for sharing the doc screenshot. It helps quite a bit to see what the formulas look like!

File: torch/nn/functional.py
Comment (nailimixaM) [reply torch/nn/functional.py:2328]: Yup, thanks!

File: torch/nn/modules/loss.py
Comment (nailimixaM) [reply torch/nn/modules/loss.py:308]: Agreed, I'll tidy this up.

**Comment by nailimixaM** [comment]
@albanD Thanks so much for those tips. How's this?

doc_new

File: torch/nn/functional.py
Comment (albanD) [inline torch/nn/functional.py:2359]: @ezyang  I suggested here that we use these 3 lines instead of doing `var = var + eps` here to reduce the impact of the epsilon for points that are far from 0.
Does that look like a good idea to you? Or you think we shouldn't do that?

**Review by albanD — COMMENTED** [review]
Thanks for the doc update, it looks great.
You might want to update it to say `max(var, eps)` instead of `var + eps` if we keep the champing.

We don't do this kind of clamping in other places so I want to get a second opinion from Ed about that. But if here is happy with it, the PR is good to go once the CI is fixed :)

You can rebase on top of master to fix the merge conflict and double check the overrides tests.

File: torch/nn/functional.py
Comment (ezyang) [reply torch/nn/functional.py:2359]: I don't have an opinion. With these sorts of stability considerations I'd prefer that we do what is usually done in the literature, but I'm not widely read enough to know what it is here.

**Comment by mruberry** [comment]
> Also did you manage to build the doc locally?

You can also see the rendered doc as an artifact of the python_doc_build. See the Developer FAQ: https://github.com/pytorch/pytorch/wiki/Developer-FAQ#if-im-editing-the-documentation-how-do-i-see-the-formatted-doc.

File: torch/nn/functional.py
Comment (jbschlosser) [reply torch/nn/functional.py:2359]: My opinion here is that I agree with @albanD's suggestion. While it is different than what e.g. `PoissonNLLLoss` does, it matches the literature.  

Quoting the [paper](https://arxiv.org/pdf/1612.01474.pdf): "We enforce the positivity constraint on the variance by passing the second output through the softplus
function, and add a minimum variance (e.g. 10^-6) for numerical stability."

The phrase "minimum variance" suggests to me that the clamping behavior is desired over an unconditional addition of `eps`.

Comment (albanD) [reply torch/nn/functional.py:2359]: Thanks for checking @jbschlosser !
So that's good and we just should update the formula to match this.

Comment (ngimel) [reply torch/nn/functional.py:2359]: There was a similar discussion about cosine similarity a year ago, went nowhere https://github.com/pytorch/pytorch/pull/31378#discussion_r362756119. 
cdist function that also has problem with occasional zero values uses clamping. 
BatchNorm and other normalizations use var + eps, in accordance with their original papers.

Comment (albanD) [reply torch/nn/functional.py:2359]: Ho I forgot about this one :D 
Thanks for the link! I guess that following the paper is the right thing to do right now.

**Comment by albanD** [comment]
Thanks for the update!
Could you rebase on top of master to make the CI run please?

File: torch/nn/functional.py
Comment (ngimel) [reply torch/nn/functional.py:2359]: I would still like us to have a consistent story for norms and this function (and cosine similarity for that matter), but it shouldn't block this PR.

**Comment by nailimixaM** [comment]
> Thanks for the update!
> Could you rebase on top of master to make the CI run please?

Sure, will do once I've just checked that those override tests are now passing. Thanks!

**Comment by nailimixaM** [comment]
@albanD I'm having issues with the pytorch_windows_vs2019_py36_cuda10.1_test1, which states that test_nn fails but gives no further details. Is there any way of getting more information/trying to fix it (long shot I know) without having to build everything from source on Windows? The tests work fine on my linux machine so I'm out of ideas 🤔 Am currently trying to build on my windows laptop which has a gpu, but it's a bit of a strain on the machine.

**Comment by ngimel** [comment]
@nailimixaM that's a known flaky test, you can ignore it.

**Comment by gfrisch** [comment]
@nailimixaM,  thank you for your work, this PR will help me a lot for my research

**Comment by nailimixaM** [comment]
@albanD I've just done a final tidy up of the documentation (+ added a note for clarity on clamping and a reference to the original paper introducing the loss function), see below. Let me know if you're happy with it and I'll rebase! 🚀 

new_doc

File: torch/nn/modules/loss.py
Comment (albanD) [inline torch/nn/modules/loss.py:358]: nit: this -> it

**Review by albanD — APPROVED** [review]
LGTM
Two small typos and this is good to go! Could you do the rebase to check CI?

**Comment by nailimixaM** [comment]
@albanD @ngimel Would you know how to fix these tests that are erroring/failing? The log of the non-flaky one (below) looks like a hardware issue, and the flaky ones are about virtual machine assignments. I can't seem to rerun any of these.

`11:34:22   test_kernel_cache_multi_gpu (__main__.TestTEFuser) ... Memory exception on virtual address 0x7f3c8ae00000, node id 5 : Page not present`
`11:34:24 GPU address 0x7f3c8ae00000, node id 4, size in byte 0x200000`
`11:34:24 Memory is allocated using hsaKmtAllocMemory`
`11:34:24 CPU address of the memory is 0x7f3c8ae00000`
`11:34:24 Memory is mapped to node id: 4 5 6 7 `
`11:34:24 Memory access fault by GPU node-5 (Agent handle: 0x5615992e1fe0) on address 0x7f3c8ae00000. Reason: Page not present or supervisor privilege.`
`11:34:24 Traceback (most recent call last):`
`11:34:24   File "test/run_test.py", line 910, in <module>`
`11:34:24     main()`
`11:34:24   File "test/run_test.py", line 889, in main`
`11:34:24     raise RuntimeError(err_message)`
`11:34:24 RuntimeError: test_jit_fuser_te failed! Received signal: SIGIOT`

**Comment by albanD** [comment]
I think you can ignore the rocm failure as it looks like a hardware failure.
Our CI on circleCI is in a bad shape since yesterday. But I restarted the runs and they should work now.

**Comment by nailimixaM** [comment]
> I think you can ignore the rocm failure as it looks like a hardware failure.
> Our CI on circleCI is in a bad shape since yesterday. But I restarted the runs and they should work now.

Looks like they're still in a bad shape 🤔

**Comment by albanD** [comment]
That should be good now. Could you rebase on top of master please?

**Comment by nailimixaM** [comment]
> That should be good now. Could you rebase on top of master please?

@albanD have I messed up here? Seem to have failing tests which I haven't written

**Comment by albanD** [comment]
I think there are some issues yes.
There are quite a lot of things in the diff that should not be here.

If you're not sure how to fix that, a simple solution for a simple PR like this  is to create a new one.
You can create a new branch from master locally. And then copy paste the relevant changes from this PR's diff page in that new branch.


As a side note, to avoid these issues, I always use rebase instead of merge to update to master. That avoid the merge commit that can be very annoying to deal with later.

**Comment by nailimixaM** [comment]
> I think there are some issues yes.
> There are quite a lot of things in the diff that should not be here.
> 
> If you're not sure how to fix that, a simple solution for a simple PR like this is to create a new one.
> You can create a new branch from master locally. And then copy paste the relevant changes from this PR's diff page in that new branch.
> 
> As a side note, to avoid these issues, I always use rebase instead of merge to update to master. That avoid the merge commit that can be very annoying to deal with later.

So sorry about that, have opened a new PR.

File: torch/nn/functional.py
Comment (nailimixaM) [reply torch/nn/functional.py:2535]: @almson See this discussion of `var` clamping

**Pull request closed without merging**
